### PR TITLE
Fix failed stress test (OpenTelemetry)

### PR DESCRIPTION
**Pull request by FrankChen021**
Fixes the stress test problem mentioned at: https://github.com/ClickHouse/ClickHouse/pull/39010#issuecomment-1236946817

The solution is to move `addAttribute` call from ctor to dtor so that both ctor and dtor are exception safe.

cc @rschu1ze @tavplubix @qoega 

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

File: src/Common/OpenTelemetryTraceContext.cpp [hunk 1]
```diff
@@ -306,6 +302,18 @@ TracingContextHolder::~TracingContextHolder()
         auto shared_span_log = current_thread_trace_context.span_log.lock();
         if (shared_span_log)
         {
+            try
```
Comment (rschu1ze) [inline src/Common/OpenTelemetryTraceContext.cpp:305]: About the ctor (l. 228ff.): Please see https://github.com/ClickHouse/ClickHouse/pull/39010#issuecomment-1238555649 about other possible exception sources. E.g. `time_since_epoch()` isn't noexcept as per standard. I know, it's a bit silly ... Maybe put a big bold try/catch around the ctor?

Comment (FrankChen021) [reply src/Common/OpenTelemetryTraceContext.cpp:305]: For this statement `std::chrono::system_clock::now().time_since_epoch()`, `now()` is noexcept, `time_since_epoch` is not declared as noexcept but it's constexpr.

But I think you're right that current ctor is exception safe based on what we do. If it changes, it's easy to lead to errors.

File: src/Common/OpenTelemetryTraceContext.cpp [hunk 2]
```diff
@@ -231,63 +247,75 @@ TracingContextHolder::TracingContextHolder(
     const Settings * settings_ptr,
     const std::weak_ptr<OpenTelemetrySpanLog> & _span_log)
 {
-    if (current_thread_trace_context.isTraceEnabled())
+    /// Use try-catch to make sure the ctor is exception safe.
+    /// If any exception is raised during the construction, the tracing is not enabled on current thread.
+    try
```
Comment (rschu1ze) [inline src/Common/OpenTelemetryTraceContext.cpp:252]: (cosmetic and just FYI for future commits: C++ has a little known feature called [function try blocks](https://en.cppreference.com/w/cpp/language/function-try-block) which would have been a good fit here)

**Review by rschu1ze — APPROVED** [review]

**Review by rschu1ze — APPROVED** [review]

File: src/Common/OpenTelemetryTraceContext.cpp (hunk 2 above)
Comment (FrankChen021) [reply src/Common/OpenTelemetryTraceContext.cpp:252]: Thanks for the information.

**Comment by FrankChen021** [comment]
There's a test in `Stateless tests (ubsan)` that still fails due to clickhouse-server does not exist. 

I didn't find any error messages related to the opentelemetry in the clickhouse-server.err.log.
Not sure what causes the failure.

**Comment by rschu1ze** [comment]
> I didn't find any error messages related to the opentelemetry in the clickhouse-server.err.log. Not sure what causes the failure.

I agree --> merging.

**Comment by azat** [comment]
>There's a test in Stateless tests (ubsan) that still fails due to clickhouse-server does not exist.

Actually it fails because of, likely, server was too slow to respond to the ping queries from `clickhouse-test`, I've seen this before.

**Comment by tavplubix** [comment]
Stateless tests (ubsan) - actually it's https://github.com/ClickHouse/ClickHouse/issues/40410

**Comment by FrankChen021** [comment]
Good to hear that, thank you for your information @tavplubix

**Pull request merged**